tests/beta/test_default_tensor.py

```diff
@@ -331,6 +338,304 @@ def test_hermitian_exceptions(self):
+        receives arguments with incompatible lengths"""
+
+        dev = tensornet_device_2_wires
+        dev._clear_network_data()
+
+        with pytest.raises(ValueError, match="must all be the same length"):
+            dev._add_initial_state_nodes(tensors, wires, names)
+
+
+    def test_add_node(self, tensornet_device_2_wires, tol):
+        """Tests that the _add_node method functions correctly."""
+
+        dev = tensornet_device_2_wires
+
+        assert len(dev._nodes["state"]) == 2
+
+        zero_state = np.array([1., 0])
+        one_qubit_gate = np.array([[0, 1], [1, 0]])
+        two_qubit_gate = np.eye(4)
+        dev._add_node(one_qubit_gate, wires=[0], name="NewNodeX")
+        dev._add_node(one_qubit_gate, wires=[1], name="NewNodeY")
+        dev._add_node(two_qubit_gate, wires=[0, 1], name="NewNodeZ")
+        assert len(dev._nodes["state"]) == 5
+        node_names = [n.name for n in dev._nodes["state"]]
+        assert set(node_names) == set(["ZeroState(0,)",
+                                       "ZeroState(1,)",
+                                       "NewNodeX(0,)",
+                                       "NewNodeY(1,)",
+                                       "NewNodeZ(0, 1)"])
+        tensors = [n.tensor for n in dev._nodes["state"]]
+        assert all([np.allclose(t.data, zero_state, atol=tol, rtol=0) for t in tensors[:2]])
+        assert all([np.allclose(t.data, one_qubit_gate, atol=tol, rtol=0) for t in tensors[2:4]])
+        assert all([np.allclose(t.data, one_qubit_gate, atol=tol, rtol=0) for t in tensors[2:4]])
+        assert np.allclose(tensors[4].data, two_qubit_gate)
+
+        assert "state" in dev._nodes and len(dev._nodes) == 1
+        dev._add_node(one_qubit_gate, wires=[0], key="junk")
+        assert "junk" in dev._nodes and len(dev._nodes) == 2
+
+
+    def test_add_edge_correctly_connects_nodes(self, tensornet_device_2_wires):
+        """Tests the _add_edge method is connecting the nodes correctly."""
+
+        dev = tensornet_device_2_wires
+
+        observable = np.array([[0, 1], [1, 0]])
+        obs_nodeA = tensornetwork.Node(observable, name="ObsA")
+        obs_nodeB = tensornetwork.Node(observable, name="ObsB")
+        dev._add_edge(obs_nodeA, 1, dev._nodes["state"][0], 0)
+        dev._add_edge(obs_nodeB, 1, dev._nodes["state"][1], 0)
+
+        # check_connected is designed to raise an exception if nodes are not connected
+        # test passes if no exception is raised
+        tensornetwork.check_connected([dev._nodes["state"][0], obs_nodeA])
+        tensornetwork.check_connected([dev._nodes["state"][1], obs_nodeB])
+
+
+    def test_create_nodes_from_tensors(self, tensornet_device_2_wires):
+        """Tests that the create_nodes_from_tensors method adds nodes to the tensor
+        network properly."""
+
+        dev = tensornet_device_2_wires
+
+        assert len(dev._nodes["state"]) == 2
+        A = np.array([[0, 1], [1, 0]])
```

Comment:
This could maybe be parametrized for other types of tensors as well (as was listed in the corresponding docstring)?

Reply:
I think I'll revisit this when we fix up `default.tensor.tf`, since we may decide to go a slightly different route (e.g., a new base class that handles the tensors in a type-agnostic way, and then separate child classes for each of the data-types)